src/e2ir.c

```diff
@@ -1875,6 +1875,68 @@ elem *toElem(Expression *e, IRState *irs)
             return e;
         }
 
+        elem *toElemBinAssign(BinAssignExp *be, int op)
+        {
+            //printf("toElemBinAssign() '%s'\n", be->toChars());
+
+            Type *tb1 = be->e1->type->toBasetype();
+            Type *tb2 = be->e2->type->toBasetype();
+
+            assert(!((tb1->ty == Tarray || tb1->ty == Tsarray ||
+                      tb2->ty == Tarray || tb2->ty == Tsarray) &&
+                     tb2->ty != Tvoid &&
+                     op != OPeq && op != OPandand && op != OPoror));
+
+            tym_t tym = totym(be->type);
+
+            elem *el;
+            elem *ev;
+            if (be->e1->op == TOKcast)
```

Comment:
No, it's intended front-end result. Although it looks problematic as AST structure, but is necessary for the dmd backend codegen.
If the assigned lhs type is narrower than rhs, CastExp will be added. For example:

```d
short a; int b;
a += b;  // Rewritten to: cast(int)a += b;
         // and it's equivalent with b = cast(short)(cast(int)a + b);
```

If we just remove the cast, backend ICE will happen, because a backend operand `OPaddass` cannot have short sized expression in its lhs.

I've tried to remove the lhs cast, but I couldn't do it.